### PR TITLE
Feature/add simulator api skeleton

### PR DESCRIPTION
**Pull request by efdx**
# Summary

In order to progress the use of simulator, implement the first revision / skeleton of the simulator ReST API. This should be usable for minimal usage of running a simulator and manipulating registers or causing modbus error responses.

This PR is made with the intention that the added API does not change the simulator, and instead tries to work on the existing base as much as possible. This is done so as not to push too many changes, making this PR intrusive or have unintended changes.

Add tests and documentation for later use and hopefully provide more detailed context to this PR.

Just to clarify, this PR in itself might not be release worthy in any ways. It is usable, but the choices made here are most likely not suitable to all use-cases. The API should be set in stone (possibly during this PR) prior to being advertised as complete.

Relates to issue #1284.

# Testing

1. Ran pytests, new tests pass and no regressions
2. The added API worked for own use-cases.

### CODE REVIEW

**Comment by janiversen** [comment]
Seems you have some pylint error, please fix.

The easiest way is to have a local development (see the README), which will run pylon locally and even fix the majority of problems automatically.

**Comment by efdx** [comment]
Will fix the CI errors.

File: pymodbus/server/simulator/http_server.py [hunk 1]
```diff
@@ -199,7 +200,7 @@ def __init__(
             "calls_json": [None, self.build_json_calls],
```
Comment (janiversen) [inline pymodbus/server/simulator/http_server.py:200]: not _json

File: pymodbus/server/simulator/http_server.py [hunk 2]
```diff
@@ -199,7 +200,7 @@ def __init__(
...
             "server_json": [None, self.build_json_server],
```
Comment (janiversen) [inline pymodbus/server/simulator/http_server.py:201]: not _json

File: pymodbus/server/simulator/http_server.py [hunk 3]
```diff
@@ -301,14 +302,18 @@ async def handle_html(self, request):
     async def handle_json(self, request):
         """Handle api registers."""
         page_type = request.path.split("/")[-1]
-        params = await request.post()
-        json_dict = self.generator_html[page_type][0].copy()
-        result = self.generator_json[page_type][1](params, json_dict)
-        return web.Response(text=f"json build: {page_type} - {params} - {result}")
+        params = await request.json()
```
Comment (janiversen) [inline pymodbus/server/simulator/http_server.py:305]: this looks very complicated, is sounds as if removing _json could simplify this a lot.

File: pymodbus/server/simulator/http_server.py [hunk 4]
```diff
@@ -301,14 +302,18 @@ async def handle_html(self, request):
...
+        json_dict = copy.deepcopy(self.generator_json[page_type][0])
```
Comment (janiversen) [inline pymodbus/server/simulator/http_server.py:306]: what happened in page_type is unknown ??

the name "page_type" does not relate well to a RestApi request, did you mean "request"

File: pymodbus/server/simulator/http_server.py [hunk 5]
```diff
@@ -460,23 +465,154 @@ def build_html_server(self, _params, html):
...
+            "function_response_clear_after": self.call_response.clear_after,
+            "call_rows": call_rows,
+            "foot": foot,
+            "result": result_txt
+        }
+
+        return json_response
 
-    def build_json_log(self, params, json_dict):
+    def build_json_log(self, _params, _json_dict):
```
Comment (janiversen) [inline pymodbus/server/simulator/http_server.py:607]: If the parameters are unused, then why demand them ?

File: pymodbus/server/simulator/http_server.py [hunk 6]
```diff
@@ -460,23 +465,154 @@ def build_html_server(self, _params, html):
...
         """Build json log page."""
-        return f"json build log: {params} - {json_dict}"
+        return {"result": "error", "error": "log endpoint not implemented"}
```
Comment (janiversen) [inline pymodbus/server/simulator/http_server.py:609]: ditto.

File: pymodbus/server/simulator/http_server.py [hunk 7]
```diff
@@ -460,23 +465,154 @@ def build_html_server(self, _params, html):
...
 
-    def build_json_server(self, params, json_dict):
+    def build_json_server(self, _params, _json_dict):
```
Comment (janiversen) [inline pymodbus/server/simulator/http_server.py:611]: ditto.

File: pyproject.toml [hunk 8]
```diff
@@ -71,6 +71,7 @@ development = [
     "pytest-profiling>=1.7.0",
     "pytest-timeout>=2.3.1",
     "pytest-xdist>=3.6.1",
+    "pytest-aiohttp>=1.0.5",
```
Comment (janiversen) [inline pyproject.toml:74]: Why is that needed now ?

it was not needed for the html ?

File: test/sub_server/test_simulator_api.py [hunk 9]
```diff
@@ -0,0 +1,467 @@
+"""Test simulator API."""
```
Comment (janiversen) [inline test/sub_server/test_simulator_api.py:1]: This is not right !! it is as if you have rewritten the whole file.

**Review by janiversen — CHANGES_REQUESTED** [review]

File: pymodbus/server/simulator/http_server.py (hunk 4 above)
Comment (efdx) [reply pymodbus/server/simulator/http_server.py:306]: > what happened in page_type is unknown ??

The aiohttp handles the unregistered endpoint. Similar to what happens when using html.

> the name "page_type" does not relate well to a RestApi request, did you mean "request"

Did not pay any mind to it. I'll change it.

File: pymodbus/server/simulator/http_server.py (hunk 3 above)
Comment (efdx) [reply pymodbus/server/simulator/http_server.py:305]: From your immediate view, I get it, and kind of agree with it. From the point of anyone automating anything based on this, not so much. But please let me know if you plan to scrap it.

Also, I think many of the things could be streamlined with changing the structure of how these are handled. IMHO the only complexity comes from the current copy-paste model I adopted, as to not break anything else, and possibly to have this discussion. Changing the logic to have the data and encoding separate, preferrably with automatic serialization with json could help alot. But I felt it would have warranted too much changes to the existing and been too intrusive.

File: test/sub_server/test_simulator_api.py (hunk 9 above)
Comment (efdx) [reply test/sub_server/test_simulator_api.py:1]: What do you mean? It's a new file. Did you

1. Confuse it with test_simulator.py?
2. Want me to combine it with it?

File: pymodbus/server/simulator/http_server.py [hunk 10]
```diff
@@ -180,7 +181,7 @@ def __init__(
         self.web_app.add_routes(
             [
                 web.get("/api/{tail:[a-z]*}", self.handle_html),
-                web.post("/api/{tail:[a-z]*}", self.handle_json),
+                web.post("/api/{tail:[a-z]*}_json", self.handle_json),
```
[the remark this answers is not in the document]
Comment (efdx) [reply pymodbus/server/simulator/http_server.py:184]: Sure. I'll change it.

File: doc/source/library/simulator/restapi.rst [hunk 11]
```diff
@@ -1,4 +1,120 @@
...
-TO BE DOCUMENTED.
+This is still a Work In Progress. There may be large changes to the API in the
+future.
+
+There is a skeleton API for the Pymodbus simulator. The API is a simple copy of
+having most of the same features as in the Web UI.
+
+The API provides the following endpoints:
+
+- /api/registers_json
```
Comment (janiversen) [inline doc/source/library/simulator/restapi.rst:12]: not _json

File: pymodbus/server/simulator/http_server.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (janiversen) [reply pymodbus/server/simulator/http_server.py:184]: Still not changed.

File: pymodbus/server/simulator/http_server.py (hunk 3 above)
Comment (janiversen) [reply pymodbus/server/simulator/http_server.py:305]: It is normal to have multiple endpoints for different types of service, so we clearly want 1 endpoint for html and another for RestAPI.

We do NOT want to use the same endpoint for different services and distinguish in the naming !

File: pymodbus/server/simulator/http_server.py (hunk 5 above)
Comment (janiversen) [reply pymodbus/server/simulator/http_server.py:607]: if a parameter is not needed, and there are no logic making the code simpler if its made dummy, then it should be removed.

File: pyproject.toml (hunk 8 above)
Comment (janiversen) [reply pyproject.toml:74]: that might be.

**Review by janiversen — CHANGES_REQUESTED** [review]
Please do not ping me, and say you changed all, when I see there are still quite a number of open comments. I do have quite a lot to do, so I will not review the same code X times, just to see its unchanged.

Once you have addressed all comments, then click the "ready for review" button.

File: test/sub_server/test_simulator_api.py (hunk 9 above)
Comment (janiversen) [reply test/sub_server/test_simulator_api.py:1]: Seems I got confused.

File: pymodbus/server/simulator/http_server.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Esco441-91) [reply pymodbus/server/simulator/http_server.py:184]: Tomorrow. Ran out of hours in my day 🙂 22:39 here.

[the remark this answers is not in the document]
Comment (Esco441-91) [reply pymodbus/server/simulator/http_server.py:184]: (Aplogies for two accounts. Github doesn't seem to like my linking)

**Comment by efdx** [comment]
> Please do not ping me, and say you changed all, when I see there are still quite a number of open comments. I do have quite a > lot to do, so I will not review the same code X times, just to see its unchanged.
>
> Once you have addressed all comments, then click the "ready for review" button.

I am very confused over what you meant. I didn't ping that the requested changes are ready, the only thing I did yesterday evening was the comments/clarification requests on the opened discussions.

Was it perhaps the prior comment related to CI changes that came in your e-mail post-fact?

image

This one was before there were any change requests posted.

Please let me know.

EDIT: Need to fix CI stuff again.

**Comment by efdx** [comment]
The CI checks now pass and I've changed the requested. Please let me know if anything else requires amending.

**Comment by janiversen** [comment]
> > Please do not ping me, and say you changed all, when I see there are still quite a number of open comments. I do have quite a > lot to do, so I will not review the same code X times, just to see its unchanged.
> > Once you have addressed all comments, then click the "ready for review" button.
> 
> I am very confused over what you meant. I didn't ping that the requested changes are ready, the only thing I did yesterday evening was the comments/clarification requests on the opened discussions.
> 
> Was it perhaps the prior comment related to CI changes that came in your e-mail post-fact?
> 
> image
> 
> This one was before there were any change requests posted.

Actually it was NOT, I read your comment and started a review, until I realized that the majority of the items was not solved.

File: pymodbus/server/simulator/http_server.py [hunk 12]
```diff
@@ -300,15 +300,18 @@ async def handle_html(self, request):
...
         """Handle api registers."""
-        page_type = request.path.split("/")[-1]
-        params = await request.post()
-        json_dict = self.generator_html[page_type][0].copy()
-        result = self.generator_json[page_type][1](params, json_dict)
-        return web.Response(text=f"json build: {page_type} - {params} - {result}")
+        endpoint = request.path.split("/")[-1]
+        params = await request.json()
+        try:
+            result = self.generator_json[endpoint](params)
```
Comment (janiversen) [inline pymodbus/server/simulator/http_server.py:306]: There are only one "endpoint" == restapi, which you defined earlier...I think you are mixing endpoints and commands.

File: pymodbus/server/simulator/http_server.py [hunk 13]
```diff
@@ -300,15 +300,18 @@ async def handle_html(self, request):
...
+        except Exception as exc: # pylint: disable=broad-exception-caught
```
Comment (janiversen) [inline pymodbus/server/simulator/http_server.py:307]: We do not like broad exceptions, especially not when it is easy replaced by a strict exception like here.

**Comment by efdx** [comment]
> Actually it was NOT, I read your comment and started a review, until I realized that the majority of the items was not solved.

I am still very confused. So for you, the post in the picture came AFTER you had reviewed my changes? If so, then I blame github for something going weird.

Regardless, I apologize if the post mentioned caused any confusion. I had no intention on pinging anyone without any changes done.

**Review by janiversen — CHANGES_REQUESTED** [review]

**Comment by janiversen** [comment]
> > Actually it was NOT, I read your comment and started a review, until I realized that the majority of the items was not solved.
> 
> I am still very confused. So for you, the post in the picture came AFTER you had reviewed my changes for the first time? If so, then I blame github for something going weird.
> 
> Regardless, I apologize if the post mentioned caused any confusion. I had no intention on pinging anyone without any changes done.

No the post came AFTER my first review (17hours ago, and you responded to that with several comments), and BEFORE my second review, I just did the THIRD review.

Anyhow its history.

File: pymodbus/server/simulator/http_server.py (hunk 13 above)
Comment (efdx) [reply pymodbus/server/simulator/http_server.py:307]: True, and my implementation is against best practices.

However, this one catches unhandled situations. Something like failed int conversion for "range_start/stop", or "value" when calling "Set". The same that aiohttp does by itself, except that it will post the data back in clear-text format, causing problems on Json parsing.

So I have few options:

1. Catch all the known exceptions in the existing code, either here on inplace with or without rethrowing as new, well-defined errors.
2. Leave it as is, handle them as unhandled
3. Combination. Catch all that are known, leave the broad exception in the event of any missed or future implementation errors.

Option 1 and 3 will require me to put in a partial improvement on json side as well, on top of everything.

Please let me know what you think.

File: pymodbus/server/simulator/http_server.py (hunk 12 above)
Comment (efdx) [reply pymodbus/server/simulator/http_server.py:306]: Yes. I often get confused with the terminology.

I'll change it to `command =`. You also had a previous suggestion on using "request" (which I missed/forgot). Let me know if you prefer that one over command.

File: pymodbus/server/simulator/http_server.py (hunk 13 above)
Comment (efdx) [reply pymodbus/server/simulator/http_server.py:307]: > on json side

on html side*

Comment (janiversen) [reply pymodbus/server/simulator/http_server.py:307]: it is easy to see which exceptions can happen, so please limit it.

Comment (efdx) [reply pymodbus/server/simulator/http_server.py]: Changed

File: pymodbus/server/simulator/http_server.py (hunk 12 above)
Comment (efdx) [reply pymodbus/server/simulator/http_server.py]: Changed

**Comment by efdx** [comment]
Apologies for the long silence. Other work stuff got in the way, but it's a poor excuse.

What do you feel like about this PR, do you still want it in? I get that long drawn-out PRs are not very attractive to be considered merging. I can live with it.

For the record, I suggest checking out [commit a1a9864](https://github.com/pymodbus-dev/pymodbus/pull/2274/commits/a1a986407f0fc90700dbb5de5d375c575dfb989f) regardless from this topic. I figure extra output would be unnecessary? The output in question can be seen in any recent CI runs that run the pytest on ubuntu.

**Comment by efdx** [comment]
I've now fixed the requested parts and put the PR up-to-date with the dev branch. Please let me know if you disagree or if there is anything else.

**Comment by abaenaDSET** [comment]
> I've now fixed the requested parts and put the PR up-to-date with the dev branch. Please let me know if you disagree or if there is anything else.

I think you should close the old changes requested if they are already solved. Then the status of the PR will become pending to merge. I see you have left one discussion opened when you answered "Changed". After this I think you will be able to mark this PR as pending to resolve.

Thanks for your contributions :)

**Comment by efdx** [comment]
> you should close the old changes requested if they are already solved. 

Thanks for the tips. I use mostly devops and gitlab internally, so I'm not that aware on all of the trick in github :)

However, I do not see a way I can mark the change request as "closed" myself. The best I can do is resolve each comment (tried), but this does not close the change request itself.

I think @janiversen needs to close them...?

**Comment by janiversen** [comment]
The review comments are closed by hitting "resolved"

**Comment by efdx** [comment]
> The review comments are closed by hitting "resolved"

The comments themselves I take it, but not the change request. Or then I'm misunderstanding something. If so, then sorry for the noise.

For me it shows everything is resolved, but it still shows "Changes requested" (and "Merging is blocked"). Same as mentioned in https://stackoverflow.com/questions/49780667/github-changes-requested-label-stay-after-i-answered-all-the-code-review-questi .

File: doc/source/library/simulator/restapi.rst [hunk 14]
```diff
@@ -1,4 +1,120 @@
 Pymodbus simulator ReST API
 ===========================
 
-TO BE DOCUMENTED.
+This is still a Work In Progress. There may be large changes to the API in the
+future.
+
+There is a skeleton API for the Pymodbus simulator. The API is a simple copy of
```
Comment (janiversen) [inline doc/source/library/simulator/restapi.rst:7]:
```suggestion
There is a skeleton API for the Pymodbus simulator. The API is a simple copy of
```
The API is a simple copy of
```

**Review by janiversen — APPROVED** [review]
Lgtm, thanks.
